### PR TITLE
Support searches for repeated annotations declared as meta-annotations on container annotations for different repeated annotations

### PR DESCRIPTION
**Pull request by gordeevnm**
## Overview

I created annotation like this:

```java
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
@Documented
@TestTemplate
@ExtendWith(DirTestCasesExtension.class)
@interface DirArgSource {
    /**
     * Describe specific files for test case args
     */
    TestCase[] value() default {};

    /**
     * Parse test cases from specified directory
     */
    String[] directory() default {};
}
```

And I want to use it with only filled `directory`, but current implementation of `org.junit.platform.commons.util.AnnotationUtils#findRepeatableAnnotations(...)` does not find `@ExtendWith` in this case

_I haven't tested this patch, it is more like demo of what i mean, but i think it must work._ 🙂

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by marcphilipp** [comment]
@gordeevnm Could you please provide a failing test case that demonstrates the issue?

**Comment by gordeevnm** [comment]
@marcphilipp something like this:
https://gist.github.com/gordeevnm/0616504bc2946c15afd230408d53aa0f

sorry, i can't provide fully worked(compilable) demo right now, so I hope the example above is informative enough

**Comment by sbrannen** [comment]
JUnit's annotation search algorithm that finds `@ExtendWith` is not affected by the attributes declared in your `@DirArgSource` annotation.

In addition, the search algorithm in question works for `@ParameterizedTest` as well as one of the test methods in the Gist you posted.

I assume it actually works for both of your test methods.

The error message you have posted in your Gist (_"You must register at least one TestTemplateInvocationContextProvider that supports @TestTemplate method"_) is generated when none of the registered `TestTemplateInvocationContextProvider`s
responds with `true` for `supportsTestTemplate(...)`.

It therefore appears that your `DirectoryArgsTestExtension` does claim to support a `@TestTemplate` method annotated with `@DirectoryArgsTestTemplate(directory = "test_case_data")`.

In light of that, I am closing this PR.

However, if you are able to create a minimal sample that reproduces the error, feel free to post back here with additional information.

**Comment by gordeevnm** [comment]
Below is a link to the project with the demo. The problem is that for the test with `@FileArgsTemplate` (in the demo project) `ext.FileArgsExtension#supportsTestTemplate` does not even run(and the constructor too)

https://github.com/gordeevnm/junit-bug-demo

**Comment by gordeevnm** [comment]
In debugger I tried to apply my patch (attached to this issue) and it worked as I want it to. I'm not sure if it's a successful patch because I understand that this annotation search algorithm is used in other places and my commit might break something. But please pay attention to this point. Thanks 🙂

**Comment by gordeevnm** [comment]
@sbrannen should i create issue instead of this PR?

**Comment by sbrannen** [comment]
Thanks for the supplying the demo project.

I'm reopening this PR so that the team can investigate if there is need for improvement in the annotation search algorithm.

Please note, however, that the recommended solution to the problem may involve a change in the way you have structured your annotations.

After taking a glance at the demo project, I noticed that the container for your repeated annotation declares attributes other than `value`.

Although that seems to be technically possible, in my experience that is nonstandard. 

When designing annotations I recommend that containers for repeatable annotations never be looked up explicitly. Their existence is only necessary for the compiler. Along those lines, I would not recommend that a container annotation declare any attributes other than `value`.

In summary, your users should never have to know about the container annotation.

If you want to allow the users to provide additional information via an annotation, it's best to create a dedicated annotation for that use case which is not a container annotation.

**Comment by sbrannen** [comment]
FYI: I have changed the title of this PR to reflect the scope of the proposal as I currently understand it.

@gordeevnm, does the new title reflect your intention?

**Comment by gordeevnm** [comment]
Yes, thank you, new title seems better. 

And I like your point about not adding extra attributes to the container annotation. You may be absolutely right, so if you decide to close this pr I won't be upset 🙂

**Comment by marcphilipp** [comment]
**Team decision:** close with the above rationale
